packages/typescript-eslint/src/config-helper.ts

```diff
@@ -92,23 +93,43 @@ export type ConfigArray = TSESLint.FlatConfig.ConfigArray;
 export function config(
   ...configs: InfiniteDepthConfigWithExtends[]
 ): ConfigArray {
-  const flattened =
+  /* We intentionally temporarily change the type to `unknown`, to ensure that we make no assumptions about the types of
+  the arguments passed to this function at runtime. If any argument is an unexpected type, we should not cause a
+  `TypeError`, and instead, we should silently return the unexpected argument unchanged, because ESLint will later
+   validate the types. */
+  const flattened: unknown[] =
     // @ts-expect-error -- intentionally an infinite type
-    configs.flat(Infinity) as ConfigWithExtends[];
+    configs.flat(Infinity);
   return flattened.flatMap((configWithExtends, configIndex) => {
-    const { extends: extendsArr, ...config } = configWithExtends;
-    if (extendsArr == null || extendsArr.length === 0) {
-      return config;
+    if (
+      !configWithExtends ||
+      typeof configWithExtends !== 'object' ||
+      !('extends' in configWithExtends)
+    ) {
+      // `configWithExtends` could be anything, but we'll assume it's a `Config` object for TS purposes.
+      return configWithExtends as Config;
+    }
+    interface ObjectWithExtends {
+      extends: InfiniteDepthConfigWithExtends[];
+    }
+    const {
+      extends: extendsArr,
+      ...config
+    }: ObjectWithExtends &
+      Partial<Record<'files' | 'ignores' | 'name', unknown>> =
+      configWithExtends as ObjectWithExtends;
+    // `config` could be any object, but we'll assume it's a `Config` object for TS purposes.
+    const configToReturn = config as Config;
```

Comment:
Inside this function, `config` should be `unknown`, but outside this function, `config` should be `Config` even if it's really an invalid value